tests/storages_tests/test_storages.py

```diff
@@ -1274,3 +1274,49 @@ def test_read_trials_from_remote_storage(storage_mode: str) -> None:
 
         study_id = storage.create_new_study()
         storage.read_trials_from_remote_storage(study_id)
+
+
+@pytest.mark.parametrize("storage_mode", STORAGE_MODES_HEARTBEAT)
+def test_retry_failed_trial_callback_repetitive_failure(storage_mode: str) -> None:
+    heartbeat_interval = 1
+    grace_period = 2
+    max_retry = 3
+    n_trials = 5
+
+    with StorageSupplier(
+        storage_mode,
+        heartbeat_interval=heartbeat_interval,
+        grace_period=grace_period,
+        failed_trial_callback=RetryFailedTrialCallback(max_retry=max_retry),
+    ) as storage:
+        study = optuna.create_study(storage=storage)
+
+        # Make repeatedly failed and retried trials by heartbeat
```

Comment:
A nit, but let me slightly fix the comment to follow the [code conventions](https://github.com/optuna/optuna/wiki/Coding-Style-Conventions#block-and-inline-comments-are-supposed-to-be-sentences).
```suggestion
        # Make repeatedly failed and retried trials by heartbeat.
```